src/compiler/type_def.rs

```diff
@@ -31,6 +31,26 @@ use crate::value::{
     Kind, Value,
 };
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Fallibility {
+    CannotFail,
+    MightFail,
+    AlwaysFails,
+}
+
+impl Fallibility {
+    #[must_use]
+    pub fn merge(&self, other: &Self) -> Self {
```

Comment:
This one needs some documentation, or potentially a more descriptive name. Without reading all the code that uses this, it's not clear what "merging" fallibility means. It might mean you have 2 expressions executed sequentially, and you are combining the fallibility of both sub-expressions (which seems to match the logic here), but there's other ways this could be merged, such as 2 parallel branches (think of an if statement, where the fallibility of each branch is different, and you want to merge, the logic will be different in that case)